### PR TITLE
Initial test around stampy.ts

### PR DESCRIPTION
**Pull request by jrhender**
This PR uses the miniflare jest env (https://miniflare.dev/testing/jest) to run a test on code that has a dependency on CF workers, in this case `server-utils/stampy.ts`. 

The new test, in `stampy.spect.ts`, relies on a data which is stored in the repository in the `mocks` directory.
The data can be refreshed manually by running `npm run refresh-test-data`.

The only app code change (other than new exports) a cut-paste of Coda URL generation  from `stampy.ts` to a separate file. Other wise, compilation of the refresh coda data script fails due to missing CF vars (e.g. `CODA_TOKEN`).

TODO:
- [x] ensure that latest code is being tested (as it seems to be the built CF worker file that is executed) 
- [x] add script for generating test data
- [x] write README on how to use the script
- [x] parameterize the test to run for each result of the mock data
- [x] ~~share the same question ids between the script and the test~~ (was too challenging to get the compilation to work)
- [x] use questionId 0 for test
- [x] manually run app to do basic sanity test (e.g. can app still get question details from Coda)

Relates to issue #191

### CODE REVIEW

File: app/mocks/coda-responses/refresh-coda-data-for-tests.ts
Comment (jrhender) [reply app/mocks/coda-responses/refresh-coda-data-for-tests.ts]: Thanks, good idea. Implemented here: https://github.com/StampyAI/stampy-ui/pull/276/commits/105e5aaae21f28075a7659af668bec048e2ed1c5

**Comment by jrhender** [comment]
Thanks for the review @Aprillion . I will make the suggested changes and then mark as Ready for Review again

File: app/server-utils/stampy.spec.ts
Comment (jrhender) [reply app/server-utils/stampy.spec.ts]: Thanks, the types are fixed! https://github.com/StampyAI/stampy-ui/pull/276/commits/5318403ac8a57ed6d6d6073cf27cc65a468e26b6

File: app/server-utils/stampy.spec.ts
Comment (jrhender) [reply app/server-utils/stampy.spec.ts]: Good suggestion, thanks. Implemented here: https://github.com/StampyAI/stampy-ui/pull/276/commits/54304f43076d99076a729a6c39191f28b1f5542a

File: app/mocks/coda-responses/refresh-coda-data-for-tests.ts
Comment (jrhender) [reply app/mocks/coda-responses/refresh-coda-data-for-tests.ts]: Good suggestion, thanks. Implemented here: https://github.com/StampyAI/stampy-ui/pull/276/commits/54304f43076d99076a729a6c39191f28b1f5542a

File: package.json
Comment (jrhender) [reply package.json:63]: Ah yes, thanks for pointing this out.

So a challenge with this setup I've proposed in this PR is that the tests need to be run on javascript that compiled for the cloudflare workers (because the `testEnvironment` has changed from `node` to `miniflare`). The tests fail if there is no `build/index.js` file. This is why `npm run build` was added to the `pr-check.yml` GH action. 

So I think I was planning to change the `test` script in this file to be `npm run dev:build && jest`. Though maybe `npm run build && jest` would be just as good.
Probably this is suboptimal and it would better if jest could compile itself immediately prior to the test execution (which is what I think `ts-jest` does if running in `node` environment). So maybe there is a way to improve in this direction 🤔 

Any thoughts on this?

File: app/server-utils/stampy.spec.ts
Comment (jrhender) [reply app/server-utils/stampy.spec.ts:24]: Ah nice, I didn't know about https://aisafety.info/questions/0 . I think that is much cleaner and more useful for assertions. I think I will change this PR to use that question instead. Though I propose that more assertions be added in a separate PR.

File: app/server-utils/stampy.spec.ts
Comment (jrhender) [reply app/server-utils/stampy.spec.ts:16]: The error message is pretty good, IMO.

I changed the `path` to be `/apis/v2` so that it doesn't match the request that `stampy.ts` makes.
The error first says:
```
TypeError: 
    >>> Error fetching https://coda.io/apis/v1/docs/fau7sl2hmG/tables/grid-sync-1059-File/rows?useColumnNames=true&sortBy=natural&valueFormat=rich&query=%22UI%20ID%22:%228486%22:
    undefined
    <<< fetch failed
```
Then after a stack trace, it says:
```
Cause:
    MockNotMatchedError: Mock dispatch not matched for path '/apis/v1/docs/fau7sl2hmG/tables/grid-sync-1059-File/rows?query=%22UI+ID%22%3A%228486%22&sortBy=natural&useColumnNames=true&valueFormat=rich': subsequent request to origin https://coda.io was not allowed (net.connect disabled)
```

To me it's pretty clear to deduce from the above that the mock path in the test needs to be changed so that is matches.

File: package.json
Comment (Aprillion) [reply package.json:63]: not sure how to make it all work seamlessly, but `npm run build && jest` sounds OK

File: app/server-utils/stampy.spec.ts
Comment (Aprillion) [reply app/server-utils/stampy.spec.ts:16]: thanks for checking, sounds good

File: package.json
Comment (jrhender) [reply package.json]: Ok, I've changed it to be `npm run build && jest`. It does seem possible to get `ts-jest` to work with miniflare, though:
https://github.com/cloudflare/miniflare-typescript-esbuild-jest/commit/bc2391a03100a15330e0a2924926fc49c7b78ee5#diff-1e058ca1442e46581b13571fb8d261f9e1f5657e26c96634d4c1072f0f0347f1 
I will perhaps give this a go in a separate PR.

Comment (jrhender) [reply package.json]: Ah, no, `npm run build && jest` doesn't quite work with the current config because `npm run build` is compiling to `public/build` where as `jest` is looking for `build/index.js` (without the `public`). So I'll have to improve this in this PR somehow.

Comment (jrhender) [reply package.json]: I'm not sure what changed, but `npm run build` now also compiling to `./build`. So `npm run build && jest` is now working just fine.

File: app/server-utils/stampy.spec.ts
Comment (jrhender) [reply app/server-utils/stampy.spec.ts:41]: Changed to question id 0 here: https://github.com/StampyAI/stampy-ui/pull/276/commits/a5d2e367d186aa704356bb884350e116d2a83490

File: app/mocks/coda-responses/refresh-coda-data-for-tests.ts
Comment (jrhender) [reply app/mocks/coda-responses/refresh-coda-data-for-tests.ts]: Thanks for the suggestion. I implemented the "extract a new method for making URL in both places" approach here: https://github.com/StampyAI/stampy-ui/pull/276/commits/e7cf3bb048fa8a933a5330d1d66f29c0977a1150

This was easier than exporting `makeCodaRequest` from `stampy.ts` because the TS compilation of `stampy.ts` expects injected CF global variables, which I thought would be tricky and confusing to do in the `refresh-coda-data-for-tests.ts` script.

**Comment by jrhender** [comment]
@Aprillion I think all of the comments from your first review have been addressed. Please give the PR another review when you have the opportunity (no rush at all 😄  ).

File: .github/workflows/pr-check.yml
Comment (jrhender) [inline .github/workflows/pr-check.yml:18]: @Aprillion does the CI have access to a `CODA_TOKEN`? If so, perhaps it would be good to refresh the test data during CI, to make sure that tests are running against the latest data from Coda. This might make the tests more flaky and unpredictable, but it would be a better test that the site will work against current data 🤔 .
```suggestion
          npm run build
          cp wrangler.toml.template wrangler.toml
          npm run refresh-test-data
          npm run test
```

Comment (Aprillion) [reply .github/workflows/pr-check.yml:18]: yes, GH actions have access to `${{ secrets.CODA_TOKEN }}` like in https://github.com/StampyAI/stampy-ui/blob/master/.github/workflows/deploy.yml

Comment (jrhender) [reply .github/workflows/pr-check.yml:18]: Ok, thanks. I propose that we run both cases during the CI workflow:
1. Run the tests against that the cached, git-committed data.
2. Refresh the data using the CODA_TOKEN and then run the tests (without committing the refresh)

In this way:
- if number 1 fails, this would suggest a regression bug in some code changes
- if number 1 passes but number 2 fails, this suggests perhaps flakiness in the coda query or that the code needs to be changed to be updated to match the latest data
- if, after a code change, number 1 fails, but number 2 passes, this could suggest that the cached data needs updated

I think that while developing, using cached data for the tests would help them run faster.

What are your thoughts on the above?

Comment (Aprillion) [reply .github/workflows/pr-check.yml:18]: hm, I like having 2 commands, so that locally I will update only once and then run the cached data

but not sure that I see the benefit of running on old data in CI, I'd be happy to only run it after refresh..

Comment (jrhender) [reply .github/workflows/pr-check.yml]: Ok, great. I changed
- the pr-check workflow to export CODA_TOKEN as an env variable and `refresh-coda-data-for-tests` to take it from the env or from the wrangler config.
- the pr-check workflow also now runs `npm run refresh-cached-data` so that the tests are run on the latest data

**Comment by jrhender** [comment]
@Aprillion The PR is ready for review again when you have time. No rush.

**Review by Aprillion — DISMISSED** [review]

**Comment by jrhender** [comment]
Hey @Aprillion, no rush, but just checking in here. What would you say the next steps for this PR would be? Are there additional reviews or changes required?
If it is indeed ready and it is a good time to merge, could you please do so when you have capacity? I don't currently have write access to the repo, so I can't merge it myself.

**Comment by Aprillion** [comment]
oh, GH actions needed some extra approval, not just PR approval.. the merge button should get enabled after the checks run I hope..

**Comment by Aprillion** [comment]
@jrhender one of the tests failed in https://github.com/StampyAI/stampy-ui/actions/runs/6245510948/job/17457463802?pr=276#step:3:139 `FAIL app/server-utils/stampy.spec.ts` -> `loadQuestionDetail › can load question 0`, can you please check?

**Comment by jrhender** [comment]
@Aprillion could you approval the workflow again when you have a moment please?
I think the issue was that the feature branch (`miniflare-jest-env`) didn't have the last changes so it was trying to make a request to Coda that probably wasn't available due to some change in Coda. I thought about merging in `master` before posting my earlier message but I was too lazy to do so 😓  .
At any rate, the CI is now passing on my GH fork (https://github.com/jrhender/stampy-ui/actions/runs/6435300979/job/17476259392?pr=2), so I'm pretty sure it will pass now in this PR.

**Comment by jrhender** [comment]
Whoops, sorry, I guess it needs to be reviewed again as well as I updated the cached coda data in the last commit: https://github.com/StampyAI/stampy-ui/pull/276/commits/30845a9c1ec090df88adefe44606facb7a60a4c8

**Review by Aprillion — APPROVED** [review]
the tests are green for me locally too, no idea why it fails in the GH actions of the main repo 🤔 

let's try to merge and see if the tests will start working for next PR - and if now, we will disable and investigate later 🤷

**Comment by Aprillion** [comment]
ok, the test passed for PR #337 => https://github.com/StampyAI/stampy-ui/actions/runs/6439654334/job/17487581099?pr=337

😌

**Pull request merged**
